### PR TITLE
[Bug] Repeated PSR

### PR DESCRIPTION
**Pull request by chMoussa**
Solves #29 and #30

### CODE REVIEW

File: horqrux/differentiation/gpsr.py [hunk 1]
```diff
@@ -232,9 +317,8 @@ def no_shots_fwd_jvp(
...
     tangent_dict = tangents[0]
 
-    # TODO: compute spectral gap through the generator which is associated with
-    # a param name.
-    spectral_gap = 2.0
+    val_keys = tuple(values.keys())
+    spectral_gap = dict.fromkeys(val_keys, 2.0)
     shift = jnp.pi / 2
 
     def jvp_component(param_name: str) -> Array:
```
Comment (RolandMacDoland) [inline horqrux/differentiation/gpsr.py:324]: Hm, it feels this can be extracted as it is reused in several places. Furthermore, I _think_ we might want to enable customization for variants of GPSR (ie. aGPSR) implementations in the future.

**Review by RolandMacDoland — COMMENTED** [review]
Nice @chMoussa. Some comments from me, mainly for the sake of discussion but also clarify some points. Happy to approve soon after.

File: horqrux/analog.py [hunk 2]
```diff
@@ -20,9 +20,15 @@ class _HamiltonianEvolution(Primitive):
...
     control: QubitSupport
+    shift: float = 0.0
+
+    def __post_init__(self) -> None:
+        super().__post_init__()
 
     def _unitary(self, values: dict[str, Array] = dict()) -> Array:
-        return expm(values["hamiltonian"] * (-1j * values["time_evolution"]))
+        # note: GPSR trick when the same param_name is used in many operations
+        time_val = values["time_evolution"] + self.shift
```
[the remark this answers is not in the document]
Comment (chMoussa) [reply horqrux/analog.py:30]: I think we have to provide the matrix directly from qadence, and use expm to perform exponentiation

File: horqrux/api.py [hunk 3]
```diff
@@ -123,25 +120,21 @@ def expectation(
...
                 state=state,
                 gates=circuit.operations,
                 observables=observables,
                 values=values,
-                n_shots=n_shots,
-                key=key,
             )
         else:
-            return no_shots_fwd(
+            chex.assert_type(n_shots, int)
```
[the remark this answers is not in the document]
Comment (chMoussa) [reply horqrux/api.py:131]: nothing in this case I believe but it replaces checkify which fails when we jit the function

File: horqrux/differentiation/gpsr.py [hunk 4]
```diff
@@ -111,9 +115,17 @@ def no_shots_fwd(
...
+        state (State): Input state or density matrix.
+        gates (Union[Primitive, Iterable[Primitive]]): Sequence of gates.
+        observables (list[Observable]): List of observables.
+        values (dict[str, float]): Parameter values.
+
+    Returns:
+        Array: Expectation values.
     """
     outputs = list(
         map(
```
[the remark this answers is not in the document]
Comment (chMoussa) [reply horqrux/differentiation/gpsr.py:131]: vmap works for batched arrays, while here we have a list of observables.

File: horqrux/differentiation/gpsr.py [hunk 5]
```diff
@@ -158,32 +180,30 @@ def align_eigenvectors(eigenvalues: Array, eigenvectors: Array) -> tuple[Array,
...
 
-def validate_permutation_matrix(P: Array) -> Array:
-    rows = P.sum(axis=0)
-    columns = P.sum(axis=1)
-    ones = jnp.ones(P.shape[0], dtype=rows.dtype)
-    return ((ones == rows) & (ones == columns)).min()
+    Returns:
+        Array: Permutation matrix P.
+    """
+    return (jnp.linalg.inv(mat) @ eigenvector_matrix).real > 0.5
```
[the remark this answers is not in the document]
Comment (chMoussa) [reply horqrux/differentiation/gpsr.py:206]: Yes, there used to be a checkify but I could not find a way to keep it with chex and jitting

File: horqrux/differentiation/gpsr.py [hunk 6]
```diff
@@ -158,32 +180,30 @@ def align_eigenvectors(eigenvalues: Array, eigenvectors: Array) -> tuple[Array,
...
 
 
 @finite_shots_fwd.defjvp
```
[the remark this answers is not in the document]
Comment (chMoussa) [reply horqrux/differentiation/gpsr.py:209]: I think with jvp, a vjp is inferred, which is why jax.grad works with it. It has to do with how you see the gradient computation as building. [Explanation here](https://docs.jax.dev/en/latest/advanced-autodiff.html#vector-jacobian-products-vjps-a-k-a-reverse-mode-autodiff).

File: horqrux/differentiation/gpsr.py [hunk 7]
```diff
@@ -158,32 +180,30 @@ def align_eigenvectors(eigenvalues: Array, eigenvectors: Array) -> tuple[Array,
 
     This is primarily used as a utility function to help sample multiple
     correlated observables when using finite shots.
-
-    Given two permuted eigenvector matrices, A and B, we wish to find a permutation
-    matrix P such that A P = B. This function calculates such a permutation
-    matrix and uses it to align each eigenvector matrix to the first eigenvector
-    matrix of eigs.
     """
     eigenvector_matrix = eigenvectors[0]
```
[the remark this answers is not in the document]
Comment (chMoussa) [reply horqrux/differentiation/gpsr.py:184]: `eigenvectors` is an array of eigenvectors, so indeed `eigenvectors[0]` is an eigenvector

File: horqrux/api.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (chMoussa) [reply horqrux/api.py]: For simplicity, I replaced with if statements and raise errors

File: horqrux/differentiation/gpsr.py (hunk 1 above)
Comment (chMoussa) [reply horqrux/differentiation/gpsr.py:321]: @RolandMacDoland I refactored matching parts of each jvp function for gpsr.

**Review by RolandMacDoland — APPROVED** [review]
Looking good ! Thanks for this @chMoussa :)

File: horqrux/differentiation/gpsr.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (chMoussa) [reply horqrux/differentiation/gpsr.py:213]: It is not a check, but it returns an array of boolean elements. I will see how to incorporate back the assert.

File: horqrux/api.py [hunk 8]
```diff
@@ -96,8 +94,7 @@ def expectation(
     observables: list[Observable],
     values: dict[str, float],
     diff_mode: DiffMode = DiffMode.AD,
-    forward_mode: ForwardMode = ForwardMode.EXACT,
-    n_shots: Optional[int] = None,
+    n_shots: int | None = None,
```
[the remark this answers is not in the document]
Comment (chMoussa) [reply horqrux/api.py:97]: We have None in pyq but indeed I can set it as int with default 0 to do no shots.
